File: clients/ios/AwareChat-iOS/AwareChat-iOS/DesignSystem/Components/MessageContainer.swift
//
//  MessageContainer.swift
//  AwareChat-iOS
//

import SwiftUI

enum ACKMessageState {
  case sending
  case sent
  case failed
}

struct MessageContainer: View {
  enum Origin {
    case sended
    case received
  }

  let origin: Origin
  let text: String
  let date: Date
  let ackState: ACKMessageState

  var body: some View {
    VStack(alignment: horizontalAlignment, spacing: Tokens.Spacing.xSmall.value) {
      BodyText(text)
        .foregroundStyle(messageTextColor)
        .padding(.horizontal, Tokens.Spacing.medium.value)
        .padding(.vertical, Tokens.Spacing.small.value)
        .background(messageBackgroundColor)
        .clipShape(RoundedRectangle(cornerRadius: Tokens.CornerRadius.medium))

      HStack(spacing: Tokens.Spacing.xSmall.value) {
        Caption2Text(date.messageTime)
          .foregroundStyle(Tokens.Colors.textSecondary)

        if let statusIcon {
          Image(systemName: statusIcon.name)
            .foregroundStyle(statusIcon.color)
            .accessibilityLabel(statusIcon.accessibilityLabel)
        }
      }
    }
    .frame(maxWidth: .infinity, alignment: frameAlignment)
  }

  private var horizontalAlignment: HorizontalAlignment {
    origin == .sended ? .trailing : .leading
  }

  private var frameAlignment: Alignment {
    origin == .sended ? .trailing : .leading
  }

  private var messageBackgroundColor: Color {
    origin == .sended ? Tokens.Colors.primary : Tokens.Colors.secondary
  }

  private var messageTextColor: Color {
    origin == .sended ? Tokens.Colors.background : Tokens.Colors.textPrimary
  }

  private var statusIcon: (name: String, color: Color, accessibilityLabel: String)? {
    guard origin == .sended else { return nil }

    switch ackState {
    case .sending:
      return nil
    case .sent:
      return (Tokens.Icons.checkmark, Tokens.Colors.primary, "Sent to server")
    case .failed:
      return (Tokens.Icons.failed, Tokens.Colors.red, "Failed to send")
    }
  }
}

#Preview("Message Containers") {
  ScrollView {
    VStack(spacing: Tokens.Spacing.medium.value) {
      MessageContainer(
        origin: .received,
        text: "Hi! Thanks for reaching out.",
        date: .now,
        ackState: .sent
      )
      
      MessageContainer(
        origin: .received,
        text: "Hi! Thanks for reaching out. And it is a reallyyyy long message to be sent by me, and is just to test how to component acts when a large message is sent",
        date: .now,
        ackState: .sent
      )

      MessageContainer(
        origin: .sended,
        text: "This message is waiting for the server.",
        date: .now,
        ackState: .sending
      )

      MessageContainer(
        origin: .sended,
        text: "This message was accepted by the server.",
        date: .now,
        ackState: .sent
      )

      MessageContainer(
        origin: .sended,
        text: "This message failed to send. And it is a reallyyyy long message to be sent by me, and is just to test how to component acts when a large message is sent",
        date: .now,
        ackState: .failed
      )
    }
    .padding(Tokens.Spacing.medium.value)
  }
  .background(Tokens.Colors.background)
}

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/designsystem/components/MessageContainer.kt
package com.example.awarechat_android.designsystem.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import com.example.awarechat_android.R
import com.example.awarechat_android.core.extensions.toMessageTime
import com.example.awarechat_android.designsystem.tokens.ColorTokens
import com.example.awarechat_android.designsystem.tokens.CornerRadiusTokens
import com.example.awarechat_android.designsystem.tokens.IconTokens
import com.example.awarechat_android.designsystem.tokens.SizeTokens
import com.example.awarechat_android.designsystem.tokens.SpacingTokens
import com.example.awarechat_android.ui.theme.AwareChatAndroidTheme
import java.time.Instant

enum class AckMessageState {
    SENDING,
    SENT,
    FAILED,
}

enum class MessageOrigin {
    SENT,
    RECEIVED,
}

@Composable
fun MessageContainer(
    origin: MessageOrigin,
    text: String,
    date: Instant,
    ackState: AckMessageState,
    modifier: Modifier = Modifier,
) {
    val isSent = origin == MessageOrigin.SENT
    val horizontalAlignment = if (isSent) {
        Alignment.End
    } else {
        Alignment.Start
    }
    val backgroundColor = if (isSent) ColorTokens.primary else ColorTokens.secondary
    val textColor = if (isSent) ColorTokens.background else ColorTokens.textPrimary

    Column(
        modifier = modifier.fillMaxWidth(),
        horizontalAlignment = horizontalAlignment,
        verticalArrangement = Arrangement.spacedBy(SpacingTokens.xSmall),
    ) {
        Surface(
            color = backgroundColor,
            shape = RoundedCornerShape(CornerRadiusTokens.medium),
        ) {
            BodyText(
                text = text,
                modifier = Modifier.padding(
                    horizontal = SpacingTokens.medium,
                    vertical = SpacingTokens.small,
                ),
                color = textColor,
            )
        }

        Row(
            horizontalArrangement = Arrangement.spacedBy(SpacingTokens.xSmall),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Caption2Text(
                text = date.toMessageTime(),
                color = ColorTokens.textSecondary,
            )

            if (isSent) {
                MessageStatusIcon(ackState = ackState)
            }
        }
    }
}

@Composable
private fun MessageStatusIcon(ackState: AckMessageState) {
    when (ackState) {
        AckMessageState.SENDING -> Unit
        AckMessageState.SENT -> Icon(
            painter = painterResource(IconTokens.checkmark),
            contentDescription = stringResource(R.string.message_sent_content_description),
            modifier = Modifier.size(SizeTokens.medium),
            tint = ColorTokens.primary,
        )
        AckMessageState.FAILED -> Icon(
            painter = painterResource(IconTokens.failed),
            contentDescription = stringResource(R.string.message_failed_content_description),
            modifier = Modifier.size(SizeTokens.medium),
            tint = ColorTokens.customRed,
        )
    }
}

@Preview(name = "Message Containers", showBackground = true)
@Composable
private fun MessageContainerPreview() {
    val previewDate = Instant.parse("2026-09-11T14:30:00Z")

    AwareChatAndroidTheme {
        Column(
            modifier = Modifier
                .verticalScroll(rememberScrollState())
                .padding(SpacingTokens.medium),
            verticalArrangement = Arrangement.spacedBy(SpacingTokens.medium),
        ) {
            MessageContainer(
                origin = MessageOrigin.RECEIVED,
                text = "Hi! Thanks for reaching out.",
                date = previewDate,
                ackState = AckMessageState.SENT,
            )
            MessageContainer(
                origin = MessageOrigin.SENT,
                text = "This message is waiting for the server.",
                date = previewDate,
                ackState = AckMessageState.SENDING,
            )
            MessageContainer(
                origin = MessageOrigin.SENT,
                text = "This message was accepted by the server.",
                date = previewDate,
                ackState = AckMessageState.SENT,
            )
            MessageContainer(
                origin = MessageOrigin.SENT,
                text = "This message failed to send. And this was a reaallyyy long messa to test how the component behavies when there's a long message inside it.",
                date = previewDate,
                ackState = AckMessageState.FAILED,
            )
        }
    }
}
